File: taurus-protect-sdk-typescript/src/services/reservation-service.ts
/**
 * Reservation service for Taurus-PROTECT SDK.
 *
 * Provides methods for managing UTXO reservations used to lock specific UTXOs
 * (Unspent Transaction Outputs) for UTXO-based blockchains like Bitcoin and
 * Litecoin, preventing double-spending during transaction creation.
 */

import { NotFoundError, ValidationError } from '../errors';
import type { ReservationsApi } from '../internal/openapi/apis/ReservationsApi';
import {
  reservationFromDto,
  reservationsFromDto,
  reservationUtxoFromDto,
} from '../mappers/reservation';
import type {
  ListReservationsOptions,
  Reservation,
  ReservationUtxo,
} from '../models/reservation';
import { BaseService } from './base';

/**
 * Service for managing UTXO reservations.
 *
 * Reservations are used to lock specific UTXOs for UTXO-based blockchains
 * like Bitcoin and Litecoin, preventing double-spending during transaction
 * creation.
 *
 * @example
 * ```typescript
 * // List all reservations
 * const reservations = await reservationService.list();
 * for (const reservation of reservations) {
 *   console.log(`${reservation.id}: ${reservation.amount}`);
 * }
 *
 * // Get a specific reservation
 * const reservation = await reservationService.get('res-123');
 * console.log(`Address: ${reservation.address}`);
 *
 * // Get UTXO details for a reservation
 * const utxo = await reservationService.getUtxo('res-123');
 * console.log(`UTXO hash: ${utxo.hash}`);
 *
 * // List reservations with filters
 * const filtered = await reservationService.list({
 *   addressId: 'addr-456',
 *   kinds: ['OUTGOING', 'CONSOLIDATION'],
 * });
 * ```
 */
export class ReservationService extends BaseService {
  private readonly reservationsApi: ReservationsApi;

  /**
   * Creates a new ReservationService instance.
   *
   * @param reservationsApi - The ReservationsApi instance from the OpenAPI client
   */
  constructor(reservationsApi: ReservationsApi) {
    super();
    this.reservationsApi = reservationsApi;
  }

  /**
   * Lists all reservations with optional filtering.
   *
   * Results are sorted by the most recent reservations first.
   * Default limit is 100 reservations.
   *
   * @param options - Optional filtering and pagination options
   * @returns Array of reservations
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List all reservations
   * const all = await reservationService.list();
   *
   * // Filter by address ID
   * const byAddress = await reservationService.list({ addressId: 'addr-123' });
   *
   * // Filter by multiple kinds
   * const byKinds = await reservationService.list({
   *   kinds: ['OUTGOING', 'CONSOLIDATION'],
   * });
   * ```
   */
  async list(options?: ListReservationsOptions): Promise<Reservation[]> {
    return this.execute(async () => {
      const response = await this.reservationsApi.walletServiceGetReservations({
        kind: options?.kind,
        address: options?.address,
        addressId: options?.addressId,
        kinds: options?.kinds,
        cursorCurrentPage: options?.cursorCurrentPage,
        cursorPageRequest: options?.cursorPageRequest,
        cursorPageSize: options?.cursorPageSize,
      });

      return reservationsFromDto(response.result);
    });
  }

  /**
   * Gets a reservation by ID.
   *
   * @param id - The reservation ID
   * @returns The reservation
   * @throws {@link ValidationError} If id is empty
   * @throws {@link NotFoundError} If reservation is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const reservation = await reservationService.get('res-123');
   * console.log(`Amount: ${reservation.amount}`);
   * console.log(`Kind: ${reservation.kind}`);
   * ```
   */
  async get(id: string): Promise<Reservation> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      const response = await this.reservationsApi.walletServiceGetReservation({
        id,
      });

      const reservation = reservationFromDto(response.result);
      if (!reservation) {
        throw new NotFoundError(`Reservation with id '${id}' not found`);
      }

      return reservation;
    });
  }

  /**
   * Gets the UTXO details for a reservation.
   *
   * Returns the UTXO (Unspent Transaction Output) associated with a reservation,
   * if any. This is relevant for UTXO-based blockchains like Bitcoin.
   *
   * @param id - The reservation ID
   * @returns The UTXO details
   * @throws {@link ValidationError} If id is empty
   * @throws {@link NotFoundError} If reservation or UTXO is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const utxo = await reservationService.getUtxo('res-123');
   * console.log(`Hash: ${utxo.hash}`);
   * console.log(`Value: ${utxo.value}`);
   * console.log(`Block height: ${utxo.blockHeight}`);
   * ```
   */
  async getUtxo(id: string): Promise<ReservationUtxo> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      const response = await this.reservationsApi.walletServiceGetReservationUTXO({
        id,
      });

      const utxo = reservationUtxoFromDto(response.result);
      if (!utxo) {
        throw new NotFoundError(`UTXO for reservation with id '${id}' not found`);
      }

      return utxo;
    });
  }
}

File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/service/ReservationService.java
package com.taurushq.sdk.protect.client.service;

import com.taurushq.sdk.protect.client.mapper.ApiExceptionMapper;
import com.taurushq.sdk.protect.client.mapper.ReservationMapper;
import com.taurushq.sdk.protect.client.model.ApiException;
import com.taurushq.sdk.protect.client.model.ApiRequestCursor;
import com.taurushq.sdk.protect.client.model.Pagination;
import com.taurushq.sdk.protect.client.model.Reservation;
import com.taurushq.sdk.protect.client.model.ReservationResult;
import com.taurushq.sdk.protect.client.model.ReservationUtxo;
import com.taurushq.sdk.protect.openapi.ApiClient;
import com.taurushq.sdk.protect.openapi.api.ReservationsApi;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetReservationReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetReservationUTXOReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetReservationsReply;

import java.util.Collections;
import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Service for managing UTXO reservations in the Taurus Protect system.
 * <p>
 * Reservations are used to lock specific UTXOs (Unspent Transaction Outputs)
 * for UTXO-based blockchains like Bitcoin and Litecoin, preventing
 * double-spending during transaction creation.
 * <p>
 * Example usage:
 * <pre>{@code
 * // First page of reservations (default page size)
 * ReservationResult reservations = client.getReservationService().getReservations();
 * // next page: getReservations(null, null, null, null, 20, reservations.getPage().getNextCursor())
 *
 * // Get a specific reservation
 * Reservation reservation = client.getReservationService().getReservation("res-123");
 *
 * // Get UTXO details for a reservation
 * ReservationUtxo utxo = client.getReservationService().getReservationUtxo("res-123");
 * }</pre>
 *
 * @see Reservation
 * @see ReservationUtxo
 */
public class ReservationService {

    private final ReservationsApi reservationsApi;
    private final ApiExceptionMapper apiExceptionMapper;
    private final ReservationMapper reservationMapper;

    /**
     * Creates a new ReservationService.
     *
     * @param apiClient          the API client for making HTTP requests
     * @param apiExceptionMapper the mapper for converting API exceptions
     * @throws NullPointerException if any parameter is null
     */
    public ReservationService(final ApiClient apiClient, final ApiExceptionMapper apiExceptionMapper) {
        checkNotNull(apiClient, "apiClient must not be null");
        checkNotNull(apiExceptionMapper, "apiExceptionMapper must not be null");
        this.reservationsApi = new ReservationsApi(apiClient);
        this.apiExceptionMapper = apiExceptionMapper;
        this.reservationMapper = ReservationMapper.INSTANCE;
    }

    /**
     * Retrieves the first page of reservations, with the default page size.
     *
     * @return the reservations and their page
     * @throws ApiException if the API call fails
     */
    public ReservationResult getReservations() throws ApiException {
        return getReservations(null, null, null, null, (ApiRequestCursor) null);
    }

    /**
     * Retrieves a page of reservations with optional filters.
     *
     * @param kind      optional kind to filter by
     * @param address   optional address to filter by
     * @param addressId optional address ID to filter by
     * @param kinds     optional list of kinds to filter by
     * @param pageSize  the page size, null or 0 for the default
     * @param cursor    a previous page's {@code getPage().getNextCursor()}, null for the first page
     * @return the reservations and their page
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if the page size is out of range
     */
    public ReservationResult getReservations(
            final String kind,
            final String address,
            final String addressId,
            final List<String> kinds,
            final Integer pageSize,
            final String cursor
    ) throws ApiException {
        return getReservations(kind, address, addressId, kinds, Pagination.page(pageSize, cursor));
    }

    /**
     * Retrieves a page of reservations with optional filters, with a low-level request cursor.
     *
     * @param kind      optional kind to filter by
     * @param address   optional address to filter by
     * @param addressId optional address ID to filter by
     * @param kinds     optional list of kinds to filter by
     * @param cursor    the request cursor, null for the first page with the default size
     * @return the reservations and their page
     * @throws ApiException if the API call fails
     */
    public ReservationResult getReservations(
            final String kind,
            final String address,
            final String addressId,
            final List<String> kinds,
            final ApiRequestCursor cursor
    ) throws ApiException {
        final CursorRequest page = CursorRequest.of(cursor);
        try {
            TgvalidatordGetReservationsReply reply = reservationsApi.walletServiceGetReservations(
                    kind, address, addressId,
                    page.currentPage(), page.pageRequest(), page.pageSizeParam(),
                    kinds);
            ReservationResult result = new ReservationResult();
            result.setReservations(reply.getResult() == null
                    ? Collections.emptyList() : reservationMapper.fromDTOList(reply.getResult()));
            return page.complete(result, PagedOperation.RESERVATIONS, reply.getCursor(), null);
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a reservation by ID.
     *
     * @param id the reservation ID
     * @return the reservation
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if id is null or empty
     */
    public Reservation getReservation(final String id) throws ApiException {
        checkArgument(id != null && !id.isEmpty(), "id must not be null or empty");
        try {
            TgvalidatordGetReservationReply reply = reservationsApi.walletServiceGetReservation(id);
            return reservationMapper.fromDTO(reply.getResult());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves the UTXO details for a reservation.
     *
     * @param id the reservation ID
     * @return the UTXO details
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if id is null or empty
     */
    public ReservationUtxo getReservationUtxo(final String id) throws ApiException {
        checkArgument(id != null && !id.isEmpty(), "id must not be null or empty");
        try {
            TgvalidatordGetReservationUTXOReply reply = reservationsApi.walletServiceGetReservationUTXO(id);
            return reservationMapper.fromUtxoDTO(reply.getResult());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }
}
